.github/workflows/ci-build.yaml

```diff
@@ -17,8 +17,99 @@ permissions:
   contents: read
 
 jobs:
+  changed-files:
+    name: Get changed files
+    outputs:
+      # reference: https://github.com/tj-actions/changed-files#outputs-
+      tests: ${{ steps.changed-files.outputs.tests_any_modified == 'true' }}
+      e2e-tests: ${{ steps.changed-files.outputs.e2e-tests_any_modified == 'true' }}
+      codegen: ${{ steps.changed-files.outputs.codegen_any_modified == 'true' }}
+      lint: ${{ steps.changed-files.outputs.lint_any_modified == 'true' }}
+      ui: ${{ steps.changed-files.outputs.ui_any_modified == 'true' }}
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 50 # assume PRs are less than 50 commits
+      - name: Get relevant files changed per group
```

Comment:
This list seems hard to maintain though

Reply:
There's no real way around that -- task runners pretty regularly ask that you keep a list of files that should be "memoized" / "cached" / "watched" as tooling does not usually track all the files that were touched.

Also we don't change tooling that frequently; these only need changing when we introduce something new. I think in the past 6+ months the only things that would have affected this list are the Kit replacement (#11928) and the `mkdocs` replacement (partly mentioned in #11390).

So I think it's actually pretty reasonable to maintain. We only really have to be careful with it when we change large parts of the build system or create a totally new directory. The vast majority of PRs do not do either of those.